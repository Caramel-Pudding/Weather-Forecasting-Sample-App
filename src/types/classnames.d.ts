declare module "classnames" {
  function classNames(...classes: string[]): string;
  export default classNames;
}
