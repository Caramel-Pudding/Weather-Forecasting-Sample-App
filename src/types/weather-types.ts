export interface City {
  id: number;
  country: string;
  name: string;
}

export interface WeatherData {
  city: City;
}
