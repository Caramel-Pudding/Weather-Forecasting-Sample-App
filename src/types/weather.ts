export enum PartOfTheDay {
  n = "night",
  d = "day",
}

export enum WeatherType {
  Clear = "Clear",
  Rain = "Rain",
  Snow = "Snow",
  Extreme = "Extreme",
}

export interface City {
  id: number;
  country: string;
  name: string;
}

export interface WeatherListItem {
  clouds: {
    all: number;
  };
  dt: number;
  dt_txt: string;
  main: {
    grnd_level: number;
    humidity: number;
    pressure: number;
    sea_level: number;
    temp: number;
    temp_kf: number;
    temp_max: number;
    temp_min: number;
  };
  rain?: {
    "3h"?: number;
  };
  snow?: {
    "3h"?: number;
  };
  sys: {
    pod: PartOfTheDay;
  };
  weather: {
    description: string;
    icon: string;
    id: number;
    main: WeatherType;
  }[];
  wind: {
    deg: number;
    speed: number;
  };
}

export interface WeatherData {
  city: City;
  cnt: number;
  cod: string;
  message: number;
  list: WeatherListItem[];
}
