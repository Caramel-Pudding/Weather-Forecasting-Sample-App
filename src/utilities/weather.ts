import { WeatherListItem } from "@/types/weather";

export const getWeatherListForToday = (
  weatherList: WeatherListItem[]
): WeatherListItem[] => {
  if (!weatherList || weatherList.length === 0) {
    return [];
  }
  const todayDate = new Date(weatherList[0].dt_txt).getDate();
  return weatherList.filter(
    (listItem) => new Date(listItem.dt_txt).getDate() === todayDate
  );
};

export const getTimeStringFromWeatherTimestamp = (
  weatherTimestamp: string
): string => {
  const timestamp = new Date(weatherTimestamp);
  return `${timestamp.getHours()}:00`;
};
