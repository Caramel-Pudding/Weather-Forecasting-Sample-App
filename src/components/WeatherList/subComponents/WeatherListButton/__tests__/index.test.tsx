import React from "react";
import { render, cleanup, fireEvent } from "@/tests/test-utils";

import { WeatherListItem } from "@/types/weather";

import { weatherDataStub } from "@/tests/stubs";
import { WeatherListButton } from "..";

describe("WeatherListButton Component", () => {
  afterEach(cleanup);

  it("Should render properly if there's proper data", () => {
    // * #TEST: ARRANGE
    const { container, getByText, getByAltText } = render(
      <WeatherListButton
        handleWeatherItemChange={() => {}}
        isSelected
        weatherItem={weatherDataStub.list[0] as WeatherListItem}
      />,
      {}
    );
    const time = getByText("15:00");
    const sunIcon = getByAltText("Clear");
    const temperature = getByText("14Â°");
    // * #TEST: ASSERT
    expect(time).toBeInTheDocument();
    expect(sunIcon).toBeInTheDocument();
    expect(temperature).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("Should conatin `containerActive` class if selected", () => {
    // * #TEST: ARRANGE
    const { container } = render(
      <WeatherListButton
        handleWeatherItemChange={() => {}}
        isSelected
        weatherItem={weatherDataStub.list[0] as WeatherListItem}
      />,
      {}
    );
    // * #TEST: ASSERT
    expect(container.firstChild).toHaveClass("containerActive");
    expect(container).toMatchSnapshot();
  });

  it("Should conatin `containerInactive` class if selected", () => {
    // * #TEST: ARRANGE
    const { container } = render(
      <WeatherListButton
        handleWeatherItemChange={() => {}}
        isSelected={false}
        weatherItem={weatherDataStub.list[0] as WeatherListItem}
      />,
      {}
    );
    // * #TEST: ASSERT
    expect(container.firstChild).toHaveClass("containerInactive");
    expect(container).toMatchSnapshot();
  });

  it("Should fire event onClick", () => {
    // * #TEST: ARRANGE
    const handleWeatherItemChange = jest.fn();
    const { container, getByRole } = render(
      <WeatherListButton
        handleWeatherItemChange={handleWeatherItemChange}
        isSelected={false}
        weatherItem={weatherDataStub.list[0] as WeatherListItem}
      />,
      {}
    );
    const button = getByRole("button");
    // * #TEST: ACT
    fireEvent.click(button as HTMLElement);
    // * #TEST: ASSERT
    expect(handleWeatherItemChange).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot();
  });

  it("Should fire event on pressing enter", () => {
    // * #TEST: ARRANGE
    const handleWeatherItemChange = jest.fn();
    const { container, getByRole } = render(
      <WeatherListButton
        handleWeatherItemChange={handleWeatherItemChange}
        isSelected={false}
        weatherItem={weatherDataStub.list[0] as WeatherListItem}
      />,
      {}
    );
    const button = getByRole("button");
    // * #TEST: ACT
    fireEvent.keyDown(button as HTMLElement, { key: "Enter", code: "Enter" });
    // * #TEST: ASSERT
    expect(handleWeatherItemChange).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot();
  });

  it("Should not fire event on pressing anything except enter", () => {
    // * #TEST: ARRANGE
    const handleWeatherItemChange = jest.fn();
    const { container, getByRole } = render(
      <WeatherListButton
        handleWeatherItemChange={handleWeatherItemChange}
        isSelected={false}
        weatherItem={weatherDataStub.list[0] as WeatherListItem}
      />,
      {}
    );
    const button = getByRole("button");
    // * #TEST: ACT
    fireEvent.keyDown(button as HTMLElement, { key: "A", code: "A" });
    // * #TEST: ASSERT
    expect(handleWeatherItemChange).toHaveBeenCalledTimes(0);
    expect(container).toMatchSnapshot();
  });
});
