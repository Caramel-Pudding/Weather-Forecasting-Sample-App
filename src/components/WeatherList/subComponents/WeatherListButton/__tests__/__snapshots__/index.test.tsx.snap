// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherListButton Component Should conatin \`containerActive\` class if selected 1`] = `
<div>
  <button
    class="container containerActive"
    type="button"
  >
    <div
      class="additionalText"
    >
      15:00
    </div>
    <section
      class="imageWrapper"
    >
      <img
        alt="Clear"
        src="icons/weather-sun.svg"
      />
    </section>
    <div
      class="temperature"
    >
      14°
    </div>
  </button>
</div>
`;

exports[`WeatherListButton Component Should conatin \`containerInactive\` class if selected 1`] = `
<div>
  <button
    class="container containerInactive"
    type="button"
  >
    <div
      class="additionalText"
    >
      15:00
    </div>
    <section
      class="imageWrapper"
    >
      <img
        alt="Clear"
        src="icons/weather-sun.svg"
      />
    </section>
    <div
      class="temperature"
    >
      14°
    </div>
  </button>
</div>
`;

exports[`WeatherListButton Component Should fire event on pressing enter 1`] = `
<div>
  <button
    class="container containerInactive"
    type="button"
  >
    <div
      class="additionalText"
    >
      15:00
    </div>
    <section
      class="imageWrapper"
    >
      <img
        alt="Clear"
        src="icons/weather-sun.svg"
      />
    </section>
    <div
      class="temperature"
    >
      14°
    </div>
  </button>
</div>
`;

exports[`WeatherListButton Component Should fire event onClick 1`] = `
<div>
  <button
    class="container containerInactive"
    type="button"
  >
    <div
      class="additionalText"
    >
      15:00
    </div>
    <section
      class="imageWrapper"
    >
      <img
        alt="Clear"
        src="icons/weather-sun.svg"
      />
    </section>
    <div
      class="temperature"
    >
      14°
    </div>
  </button>
</div>
`;

exports[`WeatherListButton Component Should not fire event on pressing anything except enter 1`] = `
<div>
  <button
    class="container containerInactive"
    type="button"
  >
    <div
      class="additionalText"
    >
      15:00
    </div>
    <section
      class="imageWrapper"
    >
      <img
        alt="Clear"
        src="icons/weather-sun.svg"
      />
    </section>
    <div
      class="temperature"
    >
      14°
    </div>
  </button>
</div>
`;

exports[`WeatherListButton Component Should render properly if there's proper data 1`] = `
<div>
  <button
    class="container containerActive"
    type="button"
  >
    <div
      class="additionalText"
    >
      15:00
    </div>
    <section
      class="imageWrapper"
    >
      <img
        alt="Clear"
        src="icons/weather-sun.svg"
      />
    </section>
    <div
      class="temperature"
    >
      14°
    </div>
  </button>
</div>
`;
